### ISSUE TITLE
Charm build can't call external dependencies

### ISSUE THREAD

**Issue by galgalesh** [juju/charm-tools#373]
## Checklist

 - [X] Confirmed this is an issue with charm-tools, not charmstore-client
 - [X] Provide versions of tools used
 - [X] Described the feature or ways to replicate the issue

## What version am I running?

I ran the following command: `snap info charm` and got the following ouput:

```
name:      charm
summary:   charm and charm-tools
publisher: charms
contact:   juju@lists.ubuntu.com
description: |
  charmstore-client and charm-tools
snap-id: 2Rryoc2ylScfbFl4eQtpntHD9iuZuMvt
commands:
  - charm
tracking:    candidate
installed:   2.2.3 (114) 103MB -
refreshed:   2017-11-02 03:23:28 +0100 CET
channels:                                   
  stable:    2.2.3                    (114) 103MB -
  candidate: 2.2.3                    (114) 103MB -
  beta:      ↑                                    
  edge:      2.2.4-dev+git-0-g716db46 (115) 103MB -
```

I am using: Ubuntu 17.10

## Issue/Feature

Charm build can't call external dependencies.

[My tactic](https://github.com/tengu-team/layer-openvpn/blob/master/tactics/download_modules.py) uses librarian-puppet to download the puppet dependencies of my Charm. `charm build` can't execute librarian-puppet.

### I expect/expected the following

charm build uses the dependencies available on the system.

### What I got

charm build can't find external dependencies.

**Comment by johnsca** [comment]
With charm-tools being strictly confined now, it can't use system dependencies.  We had to move to strict confinement due to hitting an issue with the snap Python interacting with the system Python that broke the tool entirely.  I believe there was some work in snapcraft to address that for classic snaps, so it might be possible to revisit that at some point.  Another option would be to defined a content interface in the charm snap to allow other snaps to be installed to add additional dependencies in a safe way.  Finally, I'm not at all certain, but it might be possible to have the snap support some system for the charm being built to specify deps that are installed at runtime within the confined snap environment, but that seems like it violates the immutability principle so it quite possibly won't work at all.

**Comment by galgalesh** [comment]
I tried some stuff out, and I think we should try to move back to `classic` confinement.

Some issues with strict confinement that won't be fixed in the near future:
- File permission operations are blocked: `chown`, `chmod` etc fail. These operations are crucial for packaging Puppet dependencies, for example.
- Hidden files and folders are not accessible. This for example creates the issue that you can not use a local folder as a wheelhouse override, because pip requires access to hidden files to install/package from a local folder.

If/when these issues are fixed, the content interface seems to be the best solution. That's a big if though..

Any news on whether the issues with Python in classic confinement are solved?

**Comment by johnsca** [comment]
I'm not sure I understand the comment about using a local folder as a wheelhouse override.  Do you mean using something like `git+file:///home/johnsca/charm-helpers`?  Because I was able to do that successfully just yesterday.

**Comment by galgalesh** [comment]
> I don't understand why you would need to chown a file during build

It's an internal Puppet thing. The only way you can bundle dependencies is by "installing" them to a cache directory. This is what `librarian-puppet` does. It seems that `chown` and `chmod` are necessary for installing Puppet modules because the files in the original tarballs have very dirty permissions, the files of some modules are owned by `jenkins` for example..

> I'm not sure I understand the comment about using a local folder as a wheelhouse override. Do you mean using something like git+file:///home/johnsca/charm-helpers? Because I was able to do that successfully just yesterday.

Ok, I was using `-e /path/to/charms.reactive`. Using `-e git+file:///path/to/charms.reactive#egg=charmhelpers` does the trick!

Maybe you can release two versions of charm-helpers? One confined and one not? Or would that confuse people too much?

**Comment by johnsca** [comment]
I meant to comment on this that I spoke with one of the snapcraft devs the other week and I think the consensus is that classic confinement is reasonable for dev / build tools like this, so I'm ok with moving back to classic confinement, once the Python library binding issue is fixed (which it may already be, and if not it will certainly be soon).

The only downside is that anyone who first installed the `charm` snap during the strictly confined period will not get updated to the classic version automatically.

**Comment by galgalesh** [comment]
For the update, I propose we add a final strictly-confined update to charm-tools that prints a "deprication" warning message with instructions for how to update manually, and then executes normally. Worst-case we break some CI tools that depend on specific output, but that seems better than people silently getting out of sync..

**Comment by marcoceppi** [comment]
Agreed, Cory and I spoke briefly about this last week and seems like a
sensible solution going forward.

Marco

**Comment by stub42** [comment]
Pip can fail all by itself with a strictly confined snap, as it attempts to use external tools not shipped with the snap. For example, I can no longer build one of my charms due to the following in wheelhouse.txt, which pulls a package from bzr:

```
-e bzr+lp:lazr-postgresql#egg=lazr-postgresql
```

**Comment by stub42** [comment]
... shortly followed by psycopg2, which fails with pip as PostgreSQL dependencies cannot be found
